### PR TITLE
Make data variables available in precondition context (#1085)

### LINKED ISSUE #1085: Please provide data variables in PreconditionContext
I have a test with `where`-block where one of the tests fails on Java 8 due to a JRE bug that is fixed on Java 9.
It would be great if I could ignore only the iteration that fails on Java 8 on Java 8, something like `@IgnoreIf({ jvm.java8 && (userId == 'foo') })` where `userId` is the data variable and `'foo'` its value for the bug-triggering case.

While reporting this, I observed that this also implies a request to check `IgnoreIf` per iteration, not per feature. :-)

### PR DESCRIPTION
**Pull request by Vampire**
closes #1085

---
This change is [Reviewable](https://reviewable.io/reviews/spockframework/spock/1094)

### CODE REVIEW

**Comment by leonard84** [comment]
Have a look at https://codecov.io/gh/spockframework/spock/compare/85d638823747cc26e94a2be0c4e3b86647b02418...b9e5efe970f3f21a104c9202cffe918123c22b59/diff there are several branches not covered by the tests, e.g., `specCondition` and so on.

**Comment by Vampire** [comment]
Well, I only added tests so far for the actual changes / enhancements I made.
That for example the `@IgnoreIf` on specification level is not tested is just like before.
I can of course add such tests, but for a start I wanted to show you how it could look like generally.
If you discard it altogether, there is no much sense to work on adding missing tests. :-)

File: spock-core/src/main/java/org/spockframework/runtime/extension/builtin/ConditionalExtension.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    } catch (ExtensionException ee) {
+      if (!(ee.getCause() instanceof MissingPropertyException)) {
+        throw ee;
+      }
+      MissingPropertyException mpe = (MissingPropertyException) ee.getCause();
+      if (!feature.getDataVariables().contains(mpe.getProperty())) {
+        throw ee;
+      }
+      feature.getFeatureMethod().addInterceptor(invocation -> {
+        List<String> dataVariables = invocation.getFeature().getDataVariables();
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/runtime/extension/builtin/ConditionalExtension.java]: The method is already rather complex, this should be extracted, preferably into an separate class, at least into a method

File: spock-specs/src/test/groovy/org/spockframework/smoke/extension/PendingFeatureIfExtensionSpec.groovy [hunk 2]
```diff
@@ -2,80 +2,129 @@ package org.spockframework.smoke.extension
...
 }
-    """)
+"""
 
     then:
     notThrown(AssertionError)
-    result.runCount == 1
-    result.failureCount == 0
-    result.ignoreCount == 0
+    result.testsStartedCount == 1
```
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/smoke/extension/PendingFeatureIfExtensionSpec.groovy]: You should add a `testsAbortedCount` check

File: spock-specs/src/test/groovy/org/spockframework/smoke/extension/PendingFeatureIfExtensionSpec.groovy [hunk 3]
```diff
@@ -2,80 +2,129 @@ package org.spockframework.smoke.extension
...
 
-class Foo extends Specification {
-  @PendingFeatureIf({false})
-  def bar() {
-    expect: false
+    then:
+    notThrown(AssertionError)
+    result.testsStartedCount == 1
+    result.testsFailedCount == 0
+    result.testsSkippedCount == 0
```
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/smoke/extension/PendingFeatureIfExtensionSpec.groovy]: you should check for `testsSucceededCount`

**Pull request merged**